nexus/reconfigurator/planning/src/blueprint_builder.rs

```diff
@@ -945,29 +934,144 @@ impl<'a> BlueprintZonesBuilder<'a> {
             .map(|sled_id| {
                 // Start with self.changed_zones, which contains entries for any
                 // sled whose zones config is changing in this blueprint.
-                let mut zones = self
-                    .changed_zones
-                    .remove(&sled_id)
-                    // If it's not there, use the config from the parent
-                    // blueprint.
-                    .or_else(|| self.parent_zones.get(&sled_id).cloned())
-                    // If it's not there either, then this must be a new sled
-                    // and we haven't added any zones to it yet.  Use the
+                if let Some(zones) = self.changed_zones.remove(&sled_id) {
+                    (sled_id.into_untyped_uuid(), zones.build())
+                }
+                // Next, check self.parent_zones, to represent an unchanged sled.
+                else if let Some(parent_zones) =
+                    self.parent_zones.get(&sled_id)
+                {
+                    (sled_id.into_untyped_uuid(), parent_zones.clone())
+                } else {
+                    // If the sled is not in self.parent_zones, then it must be a
+                    // new sled and we haven't added any zones to it yet.  Use the
                     // standard initial config.
-                    .unwrap_or_else(|| BlueprintZonesConfig {
-                        generation: Generation::new(),
-                        zones: vec![],
-                    });
-
-                zones.sort();
-
-                // TODO-cleanup use `TypedUuid` everywhere
-                (sled_id.into_untyped_uuid(), zones)
+                    (
+                        sled_id.into_untyped_uuid(),
+                        BlueprintZonesConfig {
+                            generation: Generation::new(),
+                            zones: vec![],
+                        },
+                    )
+                }
             })
             .collect()
     }
 }
 
+// This is a sub-module to hide implementation details from the rest of
+// blueprint_builder.
+mod builder_zones {
```

Comment:
This seemed fine reviewing this PR, but when I got to #5493, I think I'd prefer if this were in a separate file. It's really a dependency of the parent module, but when reviewing file-by-file that leaves it a little out of order. If it were a separate file, it'd be a little easier to jump to it first then review its caller. Not a big deal if you disagree.

Reply:
Hmm, while writing this code it was actually really nice to be able to make changes within the same file, but from the reviewer standpoint I can see that argument. I don't have a strong opinion, so I'll move it into its own file in #5493.